nucliadb_telemetry/nucliadb_telemetry/grpc_metrics.py

```diff
@@ -0,0 +1,99 @@
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import functools
+from typing import Any, Awaitable, Callable
+
+import grpc
+from grpc import ClientCallDetails, aio  # type: ignore
+from grpc.experimental import wrap_server_method_handler  # type: ignore
+
+from nucliadb_telemetry import metrics
+
+grpc_client_observer = metrics.Observer("grpc_client_op", labels={"method": ""})
+grpc_server_observer = metrics.Observer("grpc_server_op", labels={"method": ""})
+
+
+class MetricsServerInterceptor(aio.ServerInterceptor):
+    """
+    A GRPC server interceptor that adds metrics to the server.
+    """
+
+    async def intercept_service(
+        self,
+        continuation: Callable[
+            [grpc.HandlerCallDetails], Awaitable[grpc.RpcMethodHandler]
+        ],
+        handler_call_details: grpc.HandlerCallDetails,
+    ) -> grpc.RpcMethodHandler:
+        handler = await continuation(handler_call_details)
+        if handler and (
+            handler.request_streaming or handler.response_streaming
+        ):  # pytype: disable=attribute-error
+            return handler
+
+        def wrapper(behavior: Callable[[Any, aio.ServicerContext], Any]):
+            @functools.wraps(behavior)
+            async def wrapper(request: Any, context: aio.ServicerContext) -> Any:
+                with grpc_server_observer(
+                    labels={"method": handler_call_details.method}  # type: ignore
+                ) as observer:
+                    value = await behavior(request, context)
+                    observer.set_status(str(context.code()))
+                return value
+
+            return wrapper
+
+        if "grpc.health.v1.Health" in handler_call_details.method:  # type: ignore
+            return handler
+
+        return wrap_server_method_handler(wrapper, handler)
+
+
+def finish_metric_grpc(metric: metrics.ObserverRecorder, result):
+    code = result._cython_call._status.code()
+    metric.set_status(str(code))
+    metric.end()
+
+
+class UnaryUnaryClientInterceptor(aio.UnaryUnaryClientInterceptor):
+    async def intercept_unary_unary(
+        self, continuation, client_call_details: ClientCallDetails, request
+    ):
+        metric = grpc_client_observer(labels={"method": client_call_details.method})
+        metric.start()
+        call = await continuation(client_call_details, request)
+        call.add_done_callback(functools.partial(finish_metric_grpc, metric))
+
+        return call
+
+
+class StreamUnaryClientInterceptor(aio.StreamUnaryClientInterceptor):
+    async def intercept_stream_unary(
+        self, continuation, client_call_details: ClientCallDetails, request_iterator
+    ):
+        metric = grpc_client_observer(labels={"method": client_call_details.method})
+        call = await continuation(client_call_details, request_iterator)
+        call.add_done_callback(functools.partial(finish_metric_grpc, metric))
+
+        return call
```

Comment:
We also have UnaryStream GRPC services

Reply:
Okay, after thinking it over again, I figure we should just observe these as normal and if any additional custom instrumentation needs to be done, that can be done on a case-by-case basic.